[WIP] Dexie-based search index implementation

poltak opened the pull request:
List of all the stuff that has been/needs to be redone. No user-facing behaviour or features should change.

### Data model

Data model is laid out as Dexie models in [here](https://github.com/WorldBrain/Memex/blob/dev/dexie-search-index/src/search/search-index-new/models/index.js) and Dexie index schema is defined [here](https://github.com/WorldBrain/Memex/blob/dev/dexie-search-index/src/search/search-index-new/storage.js#L49)

- [x] Page
  - URL PK (0-1 page per URL)
  - indexes on [title/URL/content] terms (full text search)
  - index on domain (suggestions)
  - ~~missing screenshot + favicon Blobs~~
- [x] Visit
  - time + URL compound PK (N Visits per 1 Page)
  - index on URL (find all visits for page)
- [x] Bookmark
  - URL PK (0-1 per URL)
- [x] Tag
  - name + URL compound PK (0-N Tags per 1 Page)
  - index on name (suggestions)
  - index on URL (find all tags for page)

### Adding stuff

- [x] add page (+ bookmark, visits) by URL
  - main page indexing method; pages should only exist with at least a visit or bookmark
- [x] add page terms by URL
  - stage II of page visit scenario (see #268 for page visit details)
- [x] add visit interaction data by URL + time
  - stage iii/final of page visit scenario
- [x] add visit by URL
  - used for recent revisits (no page data re-indexing)

### Deleting stuff

- [x] delete pages by URLs
  - overview delete
- [x] delete pages by domain
  - opt. popup page on-blacklist-update delete hook
- [x] delete pages by pattern
  - opt. options page on-blacklist-update delete hook

### Tags specific

Simplified this interface - omitting unused methods

- [x] add tag for URL
  - tags dropdowns in popup + overview pages
- [x] delete tag for URL

### Bookmark specific

- [x] add bookmark by URL
  - via overview or popup
  - includes logic to extract data from tab if no page exists (popup-only)
- [x] delete bookmark by URL
  - unbookmarking via overview or popup
- [x] browser bookmark create listener
  - happens when user adds a bookmark via browser, not the ext.
  - may do a remote fetch (same as imports) if the page does not exist (user goes to bookmark mgmt, bookmarks a completely new page)

### Utilities

- [x] get page by URL
  - takes URL, returns `Page` model instance or `undefined`

### Imports

- [x] grab already stored history + bookmark URLs (perform set difference with browser data sources)
  - would be nice to see if we can now replace the linear in-mem diff with a Dexie query

### Search

- [x] terms search
- [x] filter by time (bm + visit events)
- [x] filter by bookmarks
- [x] filter by domain
- [x] filter by tag
- [x] blank search
- [x] blank search + filters
- [x] total (not current page) result count
- [ ] ~~result count for no-terms search (filters)~~ (not really feasible)
- [x] omnibar search working
- [x] auto-complete/suggest domains
- [x] auto-complete/suggest tags


### Known bugs

- [x] imports creates a visit for import time
- [x] ~~typing 1 letter in the popup search/tags inputs results in infinite typing loop (can't see how this is related yet...)~~ (seems to be my browser; other ext popups doing this too)
- [x] display times in overview are always the latest bookmark/visit even if time filter used (display issue only)
- [x] various imports related issues; progress sometimes isn't halted properly causing it to continue running lots of imports in background **probably not related**
- [ ] search slows down sometimes for some unknown reason; script refresh fixes it. Maybe something with Dexie
- [x] FF Promise breaks FF

### Misc TODOs

- [x] update old popup code dependent on old page structure
- [x] go over any remaining usages of Pouch
  - lots of old WebMemex parts highly coupled to Pouch
  - need to refactor within the old search index code and wrap in some index method
- [x] ensure no more usages of old page IDs exist in codebase (outside of old index dir)
  - now simply using URL (which page ID was derived from in an overly complicated way)
- [x] cleanup the Storage class and how it relates to the interface in `search-index-new/index`
- [x] ensure all the old search-index works after refactoring
  - lots of modules highly-coupled with Pouch meant big refactoring to move it into old index dir
  - get some data on older version, then update to this branch and everything should behave the same
- [x] go over everything; at least this time __a lot__ less code than prev. implementation
- [x] write some unit tests for index methods
- [x] squash this branch


Other things?

oliversauter commented:
Wow, such great work @poltak. As already mentioned, when testing it 2 days ago i could finish 15k documents in about 1.5 hours :)

But I also ran into a rather nasty bug:
#### How to reproduce:
1) Let the importer run
2) After a while it hangs itself and the ram shoots to almost 1GB and the CPU to 240%
3) Only way out is restarting the extension.In the latest test even that didn't help and I could not progress with the imports anymore. 
In order for it to work, I had to change the blacklist in order to force a recalculation. 
If you guys can't reproduce, i'd be happy to hop on a skype session and test with you what is necessary.

Also what we talked about before, does it make sense to include a term level concurrency mode, so that in any case a search request does not have to wait for a page to be indexed?

oliversauter commented:
Addition: While the importer is going, it seems like gradually increasing the RAM load, so there may be some things not properly being removed and clogging the RAM?

poltak commented:
Cool thanks for the report. Will look into. 

There's no queue used for index ops currently. Dexie seems to manage this fine, but this is another thing I should try to break and get confirmation on. Will add to the list
Dexie puts us at a much higher level than the prev IndexedDB implementation, so worrying about indexing individual terms is no longer relevant

> On Feb 28, 2018, at 16:41, Oliver Sauter <notifications@github.com> wrote:
> 
> Wow, such great work @poltak. As already mentioned, when testing it 2 days ago i could finish 15k documents in about 1.5 hours :)
> 
> But I also ran into a rather nasty bug:
> 
> How to reproduce:
> 
> Let the importer run
> After a while it hangs itself and the ram shoots to almost 1GB and the CPU to 240%
> Only way out is restarting the extension.In the latest test even that didn't help and I could not progress with the imports anymore.
> In order for it to work, I had to change the blacklist in order to force a recalculation.
> If you guys can't reproduce, i'd be happy to hop on a skype session and test with you what is necessary.
> Also what we talked about before, does it make sense to include a term level concurrency mode, so that in any case a search request does not have to wait for a page to be indexed?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

oliversauter commented:
> so worrying about indexing individual terms is no longer relevant

Does this mean Dexie handles prioritisation between search requests and indexing itself?
Just want to prevent that under high indexing load (long ass article or anything like that) we have a delay in the search speed, in the way it is now.

poltak commented:
> Does this mean Dexie handles prioritisation between search requests and indexing itself?

Dexie handles it at another level which it calls transactions. These are like a set of DB ops that either all happen or none happen (if something goes wrong, it's rolled back). In our ext, things like adding a new page or searching is a single transaction. It will handle scheduling these after each other if they are writing to the same table, but apparently allows them to run in parallel if they are just reading (search only needs to read).

That big 6k+ terms United States wiki article now only takes ~400ms for me to index the terms. Takes > 2s for me on the `master` version. Seems to be a good improvement with terms indexing. Would be good to play around with bigger DB once imports issues sorted and see how it scales

poltak commented:
Haven't looked at this yet @oliversauter. It's on the radar

poltak commented:
@oliversauter spent a bit of time playing with imports; managed to import ~2k pages pretty quickly at 20 concurrency and it stayed at fairly constant resource overhead the whole time. Didn't get into any hang or stop. We should talk about it briefly in call tonight if still a prob.

I found one issue that may be related which is sometimes it doesn't properly stop the imports progress when pause/cancel is pressed. The UI looks like it stopped fine, but in the background it's still downloading. Which isn't good as it could continue through the entire history, and the user will be left confused why the browser's going very slow. 
Not reproducible yet, just noticed it happening once, spent ages trying to get it happen again, then it happened later when I wasn't expecting. Shouldn't be related to the new index though - imports isn't really touched - probably a bug from the recent changes to imports state, but I'll see if I can get a fix in here.

oliversauter commented:
Ok, another set of clues:
The importer didn't completely hang itself anymore in the sense that the application crashed. 
But at some point it does  get stuck and only very slowly progresses. 
The only output then is a few error'd out urls, which actually should not error out.
Most of them are "data fetch failed"
<img width="762" alt="screen shot 2018-03-02 at 12 32 16" src="https://user-images.githubusercontent.com/7870039/36897392-a2c1c8aa-1e16-11e8-84f3-15cc66293262.png">

I am not sure, but maybe it has something to do with us sending too many requests?
Do we handle all HTTP cases well?

oliversauter commented:
Another round of reviews:
1) clicking on tag filter in a result element when already a search term is typed, does not work > adds the #test tag to the search, but only by clicking on the search field and adding a space, it will put the tag into the filter. 
2) When a page has ben revisited, I expect it to move to the top of the empty search. (test: import, then visit one of the pages, reload search overview after visit has been logged)
<img width="858" alt="screen shot 2018-03-03 at 16 38 21" src="https://user-images.githubusercontent.com/7870039/36936138-53b87560-1f01-11e8-88e3-46ad8b859022.png">
3) Scoring for titles and urls does not work yet. Expect the erdogan article be on top
<img width="842" alt="screen shot 2018-03-03 at 16 39 09" src="https://user-images.githubusercontent.com/7870039/36936149-6c2e2b76-1f01-11e8-92da-2a3cc89e5c8d.png">

4) sometimes when I search too fast, and it would result in an empty search result, it resorts to showing the results for no term entered. 

5) The results counter does not work properly, it shows only the number of results that are already streamed in.

6) Adding tags to pages when navigating in the same tab not possible. The first page works, everything after not. 
Test: 1) Open [this article](https://www.theguardian.com/books/2017/jun/23/why-iran-has-16-different-translations-of-one-khaled-hosseini-novel), try to tag, should work. 2) find a guardian link to another article, follow it, try to tag, tag field greyed out. In Prod it works.

poltak commented:
@oliversauter search oddities are expected until I get around to going over all that code this week and verifying things. Gonna be lots of little things to fix up. 

6 and 5 work fine for me. 6 will be greyed out if you open the popup before the initial indexing happens (DOM load + few hundred ms) - no changes there. 5 should show the correct total count as long as there are terms entered - all look correct for me with 4k pages to search through, but will keep an eye out for any wrong. Planning to look into possibly of counts for non-terms search this week

4 and 1, like imports, is not really related to these changes but will look into including fixes for them once all the other stuff is done.

One big issue to mention is the new index currently won't run under Firefox due to IndexedDB/Dexie transactions not working with FF's native `Promise` implementation. Should be able to get around it easy by updating our build to replace that code, at least for FF. This is main thing I'm playing around with, along with continuing implementation. [More info upstream](https://github.com/dfahlander/Dexie.js/issues/317).

 @oliversauter I'll let you know when I need any more manual testing or feedback just to save time and effort

ShishKabab commented:
Looks very good John, lot's of great work done here!

Regarding the testing here's some thoughts:
- Usually I try to hard code data, or write functions in the tests themselves to generate it (if those functions can be kept simple enough, since I also try to minimize logic in tests.) One example of this is createId() in runSuite(). Here maybe we break normalizeUrl() at one point, or suddenly we break the external API by changing the ID generation, and the tests won't catch that the way they're currently written. It's a debatable point, because writing this tests in this way might need more work both up front and maintaining a changing codebase, but it will probably increase the usefulness of the tests.
- We cannot assume the tests in mutation tests are executed in order. Every test() is meant to be a complete scenario. For readability, you can split up the tests in different functions, maybe with a certain naming convention, which will show up in the stack trace if something goes wrong.
- The mapResultsFunc() kind of worries me. The new and old index should have the same API right? Here I may be misunderstanding something, but to my understanding the rest of the codebase should not know if the new or old index is used.
- I'd renamed expected{1,2,3} to something more descriptive.
- The call to page.loadRels(): 1) Is this symetric with the old index API? 2) This is an implementation detail resulting from the use of Dexie, and thus should not be part of the public API. If you want to test the internals of the new index, I'd recommend moving the test inside search-index-new. Also, I'd recommend to prefix that method with an underscore, or make an entirely new object that wraps the internal object. Maybe we should have a clear policy about these things, how to hide internals and prevent people from using APIs we do not intend to support in the long run.
- As a naming convention, I usually write top-level constants like TEST_PAGE_1 instead of page1 and visit1. This makes it cleared these things are shared between multiple tests.
- What we could do as a convention is to place test data in a .test.data.js file, like I've done in the dexie-import branch. We need to decide here whether we want to use it in the tests in one of ways 1) import * as TEST_DATA from './index.test.data'; TEST_DATA.page1 or 2) import { TEST_PAGE_1, ... } from './index.test.data'. Thoughts about this?
- Not sure if this matters, but my mental model would say that visit1 < visit2 < visit3

One other question:
- When writing the import/export functionality, I noticed the code to retrieve screenshots from Pouch was gone. Will the old index keep working as expected?

oliversauter commented:
Found a bug with the domain filters. 
1) when searching for `nytimes.com` in the address bar, it shows some results that definitely not from `nytimes.com` nor do they have this term in the text.
Seems to happen for all domains entered.
<img width="1614" alt="screen shot 2018-03-08 at 20 42 58" src="https://user-images.githubusercontent.com/7870039/37174161-0dcc78ee-2316-11e8-9c79-60a6e60b80b9.png">
2) also weird is, that if i go on "show more results" it only shows one of the results:
<img width="834" alt="screen shot 2018-03-08 at 21 17 33" src="https://user-images.githubusercontent.com/7870039/37174199-2599c5a8-2316-11e8-8f44-6e2409186f6a.png">

poltak commented:
@oliversauter good find! I had totally forgotten about implementing domains filter extraction from queries (different to the UI filter). Wrote a test to confirm it indeed works on old but fails on new index. Then wrote the code to ensure it now passes. Domains search without terms seems pretty slow right now (seems linear to something).

Thanks a lot for the tests feedback @ShishKabab ! 

> ID gen

I like your points on hardcoding here instead of calling those ID deriving fns. The URL normalization stuff should get their own tests and be considered an unrelated side-effect to the actual index tests. Don't want to call them directly here. Although note that a few of the index methods still depend on them internally (anything that accepts a URL as param will transform it into an ID). Done in 
633f46c

> dependent tests

Yeah this was something I felt was a bit yucky. Have changed those mutation tests (now calling them "read-write ops tests") so that the test data gets reset before each of them, and that they don't make any assumptions about data based on other tests. Also ensured they all do assertions on results both before and after and write ops. All the outer tests now grouped under "read ops tests" which don't need to reset data each time. 35c2599

> test data

Updated names, etc. I like the idea of the separate test data modules. May make tests less clear, but I think most editors should be able to easily show what's in the imported test data for reference. Should the expected values live with test data? In this case, we only need to assert matching IDs, so the expected values are all shared. 633f46c

> `page.loadRels()` thing

Yes this was a big code smell. It is implementation specific to the Dexie Page model, and should be encapsulated in the new index method implementation methods d2dcb45 nice find

> mapResultsFunc

Yes, this was weird. Basically it's the post-search stage where constant size display data fetching for results happens,
but the input shapes differed between old and new index (output - what gets sent to UI- is always same however). Updated the old index to be in-line with the new index intermediate results shape and 
now no differences needed between the tests. 7b63ea7

> I noticed the code to retrieve screenshots from Pouch was gone. Will the old index keep working as expected?

Do you remember the weird way it was implemented before where the images were fetched from Pouch in the UI layer?
This is all encapsulated behind the index interface now, and the search result objects include optional `screenshot` and `favIcon`
`string` properties which are just the image data URIs (displayed directly). In the old index, this is now in the `map-search-to-pouch` stage
(same ops as before, just not in the UI), while in the new index, it just grabs the `Blob`s stored with `Page`s in Dexie, and serializes to data URI.

ShishKabab commented:
Looks very cool John  :)

> PouchDB attachments

Ah, missed that the code was there  :)  Have seen the code in the mapping function now, so looks good. Very happy to see that code get out of the UI layer!

> Test data

I'd say that most of the time you expect something you have put in, so it's in the data file anyway. For expected stuff that you didn't put in I'd say to leave those things either in the tests themselves or in the in the test file. Mentally it makes a the most sense for me like that.

oliversauter commented:
> good find! I had totally forgotten about implementing domains filter extraction from queries 

:) Ok. seems to be the case for #tags filter too.

oliversauter commented:
Found another search weirdness:
#### How to reproduce:
1) Search for a term1 that is not indexed yet > get no results (expected) (my example: `hammer`)
2) Search for term1 and term2 which is in a title > get a couple of results, but only the title matches (my example: `hammer time`)
3) Search only for term2 > get all results of that term, including title matches. (my example: 'time')

oliversauter commented:
Just tested the current version on Chromium and it didnt start the downloads showing the following error:
<img width="844" alt="screen shot 2018-03-14 at 18 19 14" src="https://user-images.githubusercontent.com/7870039/37419403-aea377ca-27b4-11e8-8b0f-3ddaf9b026a6.png">

poltak commented:
@oliversauter RE that error: are you sure you weren't on `dev/dexie-import` branch? This is a known build error @ShishKabab brought up with me yesterday, however it looks like a commit has been pushed to that branch to fix it now. If you're sure it was this branch (`dev/dexie-search-index`), and it's still happening, let me know. Building and installing ok here

Bit confused with the search bug you reported. step 2 seems to contradict step 1: `hammer` has results in step 2, but in step 1, it's yet to be indexed - is there a missing in-between step? What are the expected and actual results you have for each step? I can try writing a test to cover this case with some more specifics and see what needs to be fixed.

oliversauter commented:
@poltak 
Re Error: Yeah sure I am on the right branch. 
<img width="222" alt="screen shot 2018-03-15 at 08 35 34" src="https://user-images.githubusercontent.com/7870039/37449769-df05174e-282b-11e8-88ca-da9fb0f6cd2a.png">
Building and installing works, it happens when starting the import process. 

RE: bug: 
since now i had to delete this github issue page with 'hammer' in it and then tried the same saerch again, it didnt show any results anymore for 'hammer time'. Only for 'time'.
http://recordit.co/OQ44bXaby5

Also happens to me on complete reinstall and using 'rhammer time' as the word ('rhammer' is on purpose.)

ShishKabab commented:
> This is a known build error @ShishKabab brought up with me yesterday

That one actually was Symbol.asyncIterator not being defined  ;)

poltak commented:
> That one actually was Symbol.asyncIterator not being defined ;)

Yeah I was wrong. This one is saying a missing `Symbol.iterator` prop on an object, which is used by standard `for...of` loops to iterate over objects. So here some `for...of` is trying to iterate an un-iterable object. This one has me confused though. I have tried on Chrome 65 (stable), 66, and 67, and also tried corresponding Chromium builds just in case, but cannot get this one to reproduce. Also gone through all the uses of `for...of`, including the one in that error message, but cannot see any issues. I presume this is something with the changes in the build that were needed in this branch. 
@oliversauter what is the Chrome/ium version that this is happening for you on?
@ShishKabab does this happen for you? If so, which browser

RE search bug from @oliversauter:
Including an unindexed term (that isn't a stop/filtered-out word) should return 0 results regardless of any other terms in the query, going by my understanding of search. Let me know what is your expected behaviour here and I'll write a test and look to see how we can change it

oliversauter commented:
> should return 0 results regardless of any other terms in the query,

Yeah at least for now, we should stay with that behaviour.

oliversauter commented:
> what is the Chrome/ium version that this is happening for you on?

On this branch, i dont get the error: https://github.com/digi0ps/Memex/tree/search-injection/src

This is my version: https://github.com/digi0ps/Memex/tree/search-injection/src

poltak commented:
@oliversauter These are completely different branches. If reproducible for you on a particular Chrome/ium version, send me that number and I'll see if I can get it to reproduce

oliversauter commented:
> These are completely different branches.

Yeah aware of that, just giving reference to where it happens and where not, so the error might be easier to track. 

Oh I thought I posted my Chromium version last night in a separate comment. Didn't send it off. 
`Version 62.0.3202.94 (Developer Build) (64-bit)`

poltak commented:
@oliversauter Awesome, was able to reproduce with that version. I tried with @ShishKabab's `dev/dexie-imports` branch, as it contains some build fixes, and it seems to be fixed in there. We'll merge that branch into this one once it's ready, so fix is coming :)

poltak commented:
Page visits aren't logged in FF 60.0b7. The remote function call (`browser.tabs.sendMessage` call from BG script) is not getting any response from the content script (listener on `browser.runtime.onMessage` event). Code looks fine and connection being made fine in Chrome. To look into

poltak commented:
Manually tested each of the features today upgrading from an existing `master` build, following the instructions [posted recently in #dev slack channel](https://worldbrain.slack.com/archives/C89BPUM5Z/p1522301780000006). Did once on both Chrome and FF. One issue found and fixed with the return types of `grabExistingKeys` method. No other problems found.

oliversauter commented:
Oh wow, the search is a billiion lightyears faster now when using domain and tag filter. Before it took almost a minute to resolve a filter on 15k documents
Now its almost instantanously. Well done!

A couple of things that somehow don't work in this branch: 
1) Memex Google results don't show up, even though the checkmark is on (tried to (un)check it, but still does not work)
2) when i was updating the branch and reloading the extension, the import counter jumped back to 17k unprocessed documents. Adding a new blacklist item did not force a recalc, only removing it again. 
When recalced, it showed 2500 documents as not yet imported, which is weird, as It completed all of them today. I already had a couple of times the impression that something is off with the way the calculation happens, as sometimes I get different values for newly installed extensions (by 1-2k margin) 
In general I think it would be good if we had a "Recalculate Stats" button in the import overview.

poltak commented:
> Memex Google results don't show up, even though the checkmark is on (tried to (un)check it, but still does not work)

Good catch. The search response shape was changed slightly to be less "Pouch-like" in our refactoring. This needed to be updated.

The import remaining counts jumping back on update you described sounds like a problem I encountered in testing yesterday. I uncovered the cause and fixed in eaa3c94 (deduping against already stored pages wasn't working after upgrade), but I presume you would have pulled the latest commits. I tried the update process again on two different profiles today but could not reproduce the issue on either.

>  In general I think it would be good if we had a "Recalculate Stats" button in the import overview.

Good idea, especially for testing (beats messing around with the blacklist page). Wrote an initial working implementation with button next to "Start import" in 845f549

oliversauter commented:
A small bug I just found:

If there is 0 to-be-imported items, then you can start the import when selecting "include failed urls"

oliversauter commented:
Ran a few tests and it all works pretty well :)

A couple of things and then I think we can merge it. (one might be a biit bigger)
1) The notification at the end to reenable the fishing warnings (as already talked about)
2) If I pause the importer it "cancels" the few items that are currently being downloaded via XHR. > seems though as if they are redownloaded when restarting? Is that the behaviour you expect?
3) I think we need to find a better solution with the favicons for several reasons: 
- they are constantly refetched when the same domain urls are loaded (even from disk, it takes sometimes 200-300ms), i assume we can save a lot of time there. 
-> even though they dont take much space, we might abstract those away and have favicons as a separate store belonging to a certain domain and when the results are loaded we load the favicons separately. What are ideas to mitigate this?
If it is an issue to get it done quickly, how can we prevent it being a problem if we do change that later?
4) in the popup when i blacklist stuff, the red icon is gone (nothing to see there)
<img width="326" alt="screen shot 2018-04-11 at 19 28 39" src="https://user-images.githubusercontent.com/7870039/38632972-8e3bf942-3dbe-11e8-8151-1be6423f64f1.png">
5) the undo button goes to the "settings" it should go to "blacklists"

oliversauter commented:
And another thing that is still not working properly is the counters. 
They always show something different. 
I was cancelling the download with 109XX of 14XXX downloaded items. 
When I got back to the overview it still showed 14830 items. Then i pressed reload. then it showed me 8XXX and then i pressed reload again and it showed me 6XXX.
![](http://g.recordit.co/kb15aKd5u4.gif)
